File: src/main/java/com/datashift/datashift/service/ConnectionService.java
package com.datashift.datashift.service;

import com.datashift.datashift.dto.ColumnInfo;
import com.datashift.datashift.dto.ConnectionRequest;
import com.datashift.datashift.dto.ConnectionResponse;
import com.datashift.datashift.dto.ConnectionTestResponse;
import com.datashift.datashift.dto.TableSchema;
import com.datashift.datashift.model.Connection;
import com.datashift.datashift.repository.ConnectionRepository;
import jakarta.persistence.EntityNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Service
@Transactional
public class ConnectionService {

    private final ConnectionRepository connectionRepository;
    private final EncryptionService encryptionService;

    public ConnectionService(ConnectionRepository connectionRepository, EncryptionService encryptionService) {
        this.connectionRepository = connectionRepository;
        this.encryptionService = encryptionService;
    }

    public ConnectionResponse createConnection(ConnectionRequest request) {
        if (connectionRepository.findByName(request.getName()).isPresent()) {
            throw new IllegalArgumentException("Connection with name '" + request.getName() + "' already exists");
        }

        String encryptedPassword = encryptionService.encrypt(request.getPassword());

        Connection connection = Connection.builder()
                .name(request.getName())
                .type(request.getType())
                .host(request.getHost())
                .port(request.getPort())
                .username(request.getUsername())
                .password(encryptedPassword)
                .databaseName(request.getDatabaseName())
                .build();

        Connection saved = connectionRepository.save(connection);
        return mapToResponse(saved);
    }

    public List<ConnectionResponse> getAllConnections() {
        return connectionRepository.findAll().stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    public ConnectionResponse getConnectionById(Long id) {
        Connection connection = connectionRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Connection not found with id: " + id));
        return mapToResponse(connection);
    }

    public void deleteConnection(Long id) {
        if (!connectionRepository.existsById(id)) {
            throw new EntityNotFoundException("Connection not found with id: " + id);
        }
        connectionRepository.deleteById(id);
    }

    public ConnectionTestResponse testConnection(Long id) {
        Connection connection = connectionRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Connection not found with id: " + id));

        String decryptedPassword = encryptionService.decrypt(connection.getPassword());
        String jdbcUrl = buildJdbcUrl(connection);

        try (java.sql.Connection dbConnection = DriverManager.getConnection(jdbcUrl, connection.getUsername(), decryptedPassword)) {
            return ConnectionTestResponse.builder()
                    .success(true)
                    .message("Connection successful")
                    .build();
        } catch (Exception e) {
            return ConnectionTestResponse.builder()
                    .success(false)
                    .message("Connection failed: " + e.getMessage())
                    .build();
        }
    }

    private String buildJdbcUrl(Connection conn) {
        String type = conn.getType().toUpperCase();
        String host = conn.getHost();
        int port = conn.getPort();
        String dbName = conn.getDatabaseName() != null && !conn.getDatabaseName().isEmpty() ? conn.getDatabaseName() : "test";

        return switch (type) {
            case "H2" -> {
                String h = host.toLowerCase();
                if (h.contains("mem")) {
                    yield "jdbc:h2:mem:" + dbName;
                } else if (h.contains("file") || host.startsWith("/") || host.startsWith(".") || port <= 0) {
                    // Support file-based H2: e.g., jdbc:h2:file:/path/to/db or jdbc:h2:file:./data/db
                    String filePath = host.contains(":") ? host : (host.startsWith("/") ? host : "./" + host);
                    yield "jdbc:h2:file:" + filePath + (dbName.equals("test") ? "" : "/" + dbName);
                } else {
                    yield "jdbc:h2:tcp://" + host + ":" + port + "/" + dbName;
                }
            }
            case "MYSQL", "MARIADB" -> "jdbc:mysql://" + host + ":" + port + "/" + dbName;
            case "POSTGRESQL", "POSTGRES" -> "jdbc:postgresql://" + host + ":" + port + "/" + dbName;
            default -> "jdbc:" + type.toLowerCase() + "://" + host + ":" + port + "/" + dbName;
        };
    }

    public List<TableSchema> getSchemas(Long id) {
        Connection connection = connectionRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Connection not found with id: " + id));

        String decryptedPassword = encryptionService.decrypt(connection.getPassword());
        String jdbcUrl = buildJdbcUrl(connection);

        List<TableSchema> schemas = new ArrayList<>();
        try (java.sql.Connection dbConnection = DriverManager.getConnection(jdbcUrl, connection.getUsername(), decryptedPassword)) {
            DatabaseMetaData metaData = dbConnection.getMetaData();
            ResultSet tables = metaData.getTables(null, null, "%", new String[]{"TABLE"});

            while (tables.next()) {
                String tableName = tables.getString("TABLE_NAME");
                List<ColumnInfo> columns = new ArrayList<>();

                ResultSet cols = metaData.getColumns(null, null, tableName, "%");
                while (cols.next()) {
                    String colName = cols.getString("COLUMN_NAME");
                    String colType = cols.getString("TYPE_NAME");
                    boolean nullable = cols.getInt("NULLABLE") != DatabaseMetaData.columnNoNulls;
                    columns.add(ColumnInfo.builder()
                            .name(colName)
                            .type(colType)
                            .nullable(nullable)
                            .build());
                }
                cols.close();

                schemas.add(TableSchema.builder()
                        .name(tableName)
                        .columns(columns)
                        .build());
            }
            tables.close();
        } catch (Exception e) {
            throw new RuntimeException("Failed to fetch schema: " + e.getMessage(), e);
        }
        return schemas;
    }

    private ConnectionResponse mapToResponse(Connection connection) {
        return ConnectionResponse.builder()
                .id(connection.getId())
                .name(connection.getName())
                .type(connection.getType())
                .host(connection.getHost())
                .port(connection.getPort())
                .username(connection.getUsername())
                .databaseName(connection.getDatabaseName())
                .createdAt(connection.getCreatedAt())
                .updatedAt(connection.getUpdatedAt())
                .build();
    }
}

File: python/app/services/connection_service.py
import time
import asyncio
from typing import Sequence
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.connection import Connection
from app.schemas.connection import (
    ConnectionRequest,
    ConnectionResponse,
    ConnectionTestResponse,
    TableSchema,
    ColumnInfo,
)
from app.services.encryption_service import encryption_service


class ConnectionService:
    def __init__(self, db: AsyncSession):
        self.db = db
    
    async def create(self, request: ConnectionRequest) -> ConnectionResponse:
        existing = await self.db.execute(
            select(Connection).where(Connection.name == request.name)
        )
        if existing.scalar_one_or_none():
            raise ValueError(f"Connection with name '{request.name}' already exists")
        
        connection = Connection(
            name=request.name,
            type=request.type,
            host=request.host,
            port=request.port,
            database_name=request.database_name,
            username=request.username,
            password=encryption_service.encrypt(request.password),
        )
        
        self.db.add(connection)
        await self.db.commit()
        await self.db.refresh(connection)
        
        return self._to_response(connection)
    
    async def get_all(self) -> list[ConnectionResponse]:
        result = await self.db.execute(select(Connection))
        connections = result.scalars().all()
        return [self._to_response(c) for c in connections]
    
    async def get_by_id(self, connection_id: int) -> ConnectionResponse:
        connection = await self._get_connection(connection_id)
        return self._to_response(connection)
    
    async def delete(self, connection_id: int) -> None:
        connection = await self._get_connection(connection_id)
        await self.db.delete(connection)
        await self.db.commit()
    
    async def test_connection(self, connection_id: int) -> ConnectionTestResponse:
        connection = await self._get_connection(connection_id)
        decrypted_password = encryption_service.decrypt(connection.password)
        
        start_time = time.time()
        try:
            conn = await self._create_db_connection(
                connection.type,
                connection.host,
                connection.port,
                connection.database_name,
                connection.username,
                decrypted_password,
            )
            if conn:
                await self._close_db_connection(conn, connection.type)
            
            elapsed_ms = int((time.time() - start_time) * 1000)
            return ConnectionTestResponse(
                success=True,
                message="Connection successful",
                connection_time_ms=elapsed_ms,
            )
        except Exception as e:
            elapsed_ms = int((time.time() - start_time) * 1000)
            return ConnectionTestResponse(
                success=False,
                message=str(e),
                connection_time_ms=elapsed_ms,
            )
    
    async def get_schemas(self, connection_id: int) -> list[TableSchema]:
        connection = await self._get_connection(connection_id)
        decrypted_password = encryption_service.decrypt(connection.password)
        
        try:
            schemas = await self._fetch_schemas(
                connection.type,
                connection.host,
                connection.port,
                connection.database_name,
                connection.username,
                decrypted_password,
            )
            return schemas
        except Exception as e:
            raise ValueError(f"Failed to fetch schemas: {e}")
    
    async def _get_connection(self, connection_id: int) -> Connection:
        result = await self.db.execute(
            select(Connection).where(Connection.id == connection_id)
        )
        connection = result.scalar_one_or_none()
        if not connection:
            raise ValueError(f"Connection with id {connection_id} not found")
        return connection
    
    async def get_connection_entity(self, connection_id: int) -> Connection:
        return await self._get_connection(connection_id)
    
    def _to_response(self, connection: Connection) -> ConnectionResponse:
        return ConnectionResponse(
            id=connection.id,
            name=connection.name,
            type=connection.type,
            host=connection.host,
            port=connection.port,
            database_name=connection.database_name,
            username=connection.username,
            created_at=connection.created_at,
            updated_at=connection.updated_at,
        )
    
    def build_jdbc_url(self, conn: Connection) -> str:
        db_type = conn.type.lower()
        if db_type == "h2":
            if "mem:" in conn.database_name:
                return f"jdbc:h2:mem:{conn.database_name.replace('mem:', '')}"
            elif "tcp:" in conn.database_name or conn.host:
                return f"jdbc:h2:tcp://{conn.host}:{conn.port}/{conn.database_name}"
            else:
                return f"jdbc:h2:file:{conn.database_name}"
        elif db_type in ("mysql", "mariadb"):
            return f"jdbc:mysql://{conn.host}:{conn.port}/{conn.database_name}"
        elif db_type == "postgresql":
            return f"jdbc:postgresql://{conn.host}:{conn.port}/{conn.database_name}"
        else:
            return f"jdbc:{db_type}://{conn.host}:{conn.port}/{conn.database_name}"
    
    def build_python_url(self, conn: Connection, decrypted_password: str) -> str:
        db_type = conn.type.lower()
        if db_type == "sqlite":
            return f"sqlite:///{conn.database_name}"
        elif db_type in ("mysql", "mariadb"):
            return f"mysql+pymysql://{conn.username}:{decrypted_password}@{conn.host}:{conn.port}/{conn.database_name}"
        elif db_type == "postgresql":
            return f"postgresql+psycopg2://{conn.username}:{decrypted_password}@{conn.host}:{conn.port}/{conn.database_name}"
        elif db_type == "h2":
            return f"h2://{conn.host}:{conn.port}/{conn.database_name}"
        else:
            return f"{db_type}://{conn.username}:{decrypted_password}@{conn.host}:{conn.port}/{conn.database_name}"
    
    async def _create_db_connection(
        self,
        db_type: str,
        host: str,
        port: int,
        database: str,
        username: str,
        password: str,
    ):
        db_type = db_type.lower()
        
        if db_type == "sqlite":
            import aiosqlite
            return await aiosqlite.connect(database)
        elif db_type in ("mysql", "mariadb"):
            import aiomysql
            return await aiomysql.connect(
                host=host,
                port=port,
                user=username,
                password=password,
                db=database,
            )
        elif db_type == "postgresql":
            import asyncpg
            return await asyncpg.connect(
                host=host,
                port=port,
                user=username,
                password=password,
                database=database,
            )
        else:
            raise ValueError(f"Unsupported database type: {db_type}")
    
    async def _close_db_connection(self, conn, db_type: str):
        db_type = db_type.lower()
        if db_type == "sqlite":
            await conn.close()
        elif db_type in ("mysql", "mariadb"):
            conn.close()
        elif db_type == "postgresql":
            await conn.close()
    
    async def _fetch_schemas(
        self,
        db_type: str,
        host: str,
        port: int,
        database: str,
        username: str,
        password: str,
    ) -> list[TableSchema]:
        db_type = db_type.lower()
        schemas = []
        
        if db_type == "sqlite":
            import aiosqlite
            async with aiosqlite.connect(database) as conn:
                cursor = await conn.execute(
                    "SELECT name FROM sqlite_master WHERE type='table'"
                )
                tables = await cursor.fetchall()
                for (table_name,) in tables:
                    cursor = await conn.execute(f"PRAGMA table_info({table_name})")
                    columns_info = await cursor.fetchall()
                    columns = [
                        ColumnInfo(
                            name=col[1],
                            type=col[2],
                            nullable=not col[3],
                            size=None,
                        )
                        for col in columns_info
                    ]
                    schemas.append(TableSchema(table_name=table_name, columns=columns))
        
        elif db_type in ("mysql", "mariadb"):
            import aiomysql
            conn = await aiomysql.connect(
                host=host, port=port, user=username, password=password, db=database
            )
            try:
                async with conn.cursor() as cursor:
                    await cursor.execute("SHOW TABLES")
                    tables = await cursor.fetchall()
                    for (table_name,) in tables:
                        await cursor.execute(f"DESCRIBE {table_name}")
                        columns_info = await cursor.fetchall()
                        columns = [
                            ColumnInfo(
                                name=col[0],
                                type=col[1],
                                nullable=col[2] == "YES",
                                size=None,
                            )
                            for col in columns_info
                        ]
                        schemas.append(TableSchema(table_name=table_name, columns=columns))
            finally:
                conn.close()
        
        elif db_type == "postgresql":
            import asyncpg
            conn = await asyncpg.connect(
                host=host, port=port, user=username, password=password, database=database
            )
            try:
                tables = await conn.fetch(
                    """
                    SELECT table_name FROM information_schema.tables 
                    WHERE table_schema = 'public' AND table_type = 'BASE TABLE'
                    """
                )
                for table in tables:
                    table_name = table["table_name"]
                    columns_info = await conn.fetch(
                        """
                        SELECT column_name, data_type, is_nullable, character_maximum_length
                        FROM information_schema.columns
                        WHERE table_name = $1
                        ORDER BY ordinal_position
                        """,
                        table_name,
                    )
                    columns = [
                        ColumnInfo(
                            name=col["column_name"],
                            type=col["data_type"],
                            nullable=col["is_nullable"] == "YES",
                            size=col["character_maximum_length"],
                        )
                        for col in columns_info
                    ]
                    schemas.append(TableSchema(table_name=table_name, columns=columns))
            finally:
                await conn.close()
        
        return schemas
